torch/distributed/fsdp/_wrap_utils.py

```diff
@@ -0,0 +1,125 @@
+import collections
+import functools
+import warnings
+from typing import Any, Callable, Deque, Dict, List, Set
+
+import torch.nn as nn
+from torch.distributed.fsdp._utils import (
+    _contains_batchnorm,
+    _override_batchnorm_mixed_precision,
+)
+from torch.distributed.fsdp.wrap import (
+    _or_policy,
+    _recursive_wrap,
+    _wrap_batchnorm_individually,
+)
+
+
+def _auto_wrap(
+    auto_wrap_kwargs: Dict[str, Any],
+    fsdp_kwargs: Dict[str, Any],
+    module_wrapper_cls: Any,  # e.g. `FullyShardedDataParallel`
+) -> None:
+    """
+    Recursively auto wraps the root module given by the key "module" in
+    ``auto_wrap_kwargs`` with the arguments in ``auto_wrap_kwargs`` and
+    ``fsdp_kwargs``.
+
+    Precondition: ``auto_wrap_policy`` contains the arguments expected by
+    ``_recursive_wrap()``, where ``auto_wrap_policy`` is not ``None``.
+    ``fsdp_kwargs`` contains all FSDP arguments except ``module``.
+    """
+    auto_wrap_policy = auto_wrap_kwargs["auto_wrap_policy"]
+    root_module = auto_wrap_kwargs["module"]
+    assert auto_wrap_policy is not None
```

Comment:
can be done in followups. Any reason only assert on `auto_wrap_policy`, but not `root_module`

Reply:
I think this was because `auto_wrap_policy` is defaulted to `None` in the FSDP constructor, so this `assert auto_wrap_policy is not None` is reflecting the intention that this function should only be run if the user specified an auto wrap policy.